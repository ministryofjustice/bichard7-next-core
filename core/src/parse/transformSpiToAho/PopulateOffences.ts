import { lookupAlcoholLevelMethodBySpiCode, lookupResultQualifierCodeByCjsCode } from "../../dataLookup"
import { COMMON_LAWS, INDICTMENT } from "../../lib/offenceTypes"
import resultCodeIsOnStopList from "../../lib/result/resultCodeIsOnStopList"
import type { CriminalProsecutionReference, Offence, OffenceCode } from "../../types/AnnotatedHearingOutcome"
import type { OffenceParsedXml, ResultedCaseMessageParsedXml, SpiResult } from "../../types/SpiResult"
import PopulateOffenceResults from "./PopulateOffenceResults"
import removeSeconds from "./removeSeconds"

const enteredInErrorResultCode = 4583 // Hearing Removed
const dontKnowValue = "D"
const adjournmentSineDieResultCode = 2007
// TODO: Refactor into an enum
const timeRange = {
  ON_OR_IN: 1,
  BEFORE: 2,
  AFTER: 3,
  BETWEEN: 4,
  ON_OR_ABOUT: 5,
  ON_OR_BEFORE: 6
}

export interface OffencesResult {
  offences: Offence[]
  bailConditions: string[]
}

const adjournmentSineDieConditionMet = (spiResults: SpiResult[]) => {
  let a2007ResultFound = false
  let aFailConditionResultFound = false

  spiResults.forEach((result) => {
    const resultCode = result.ResultCode !== undefined ? Number(result.ResultCode) : undefined
    if (resultCode === adjournmentSineDieResultCode) {
      a2007ResultFound = true
    } else if (!resultCodeIsOnStopList(resultCode ?? 1000)) {
      aFailConditionResultFound = true
    }
  })

  return a2007ResultFound && !aFailConditionResultFound
}

export default class {
  private bailConditions: string[] = []

  constructor(
    private courtResult: ResultedCaseMessageParsedXml,
    private hearingDefendantBailConditions: string[] = [],
    private isAdjournmentSineDieConditionMet = false
  ) {}

  private getOffenceReason = (spiOffenceCode: string): OffenceCode => {
    const spiOffenceCodeLength = spiOffenceCode.length
    const offenceCode: Partial<OffenceCode> = {
      Reason: spiOffenceCodeLength > 4 ? spiOffenceCode.substring(4, Math.min(7, spiOffenceCodeLength)) : "",
      ...(spiOffenceCodeLength > 7 ? { Qualifier: spiOffenceCode.substring(7) } : {}),
      FullCode: spiOffenceCode
    }

    if (spiOffenceCode.startsWith(COMMON_LAWS)) {
      offenceCode.__type = "CommonLawOffenceCode"
      if (offenceCode.__type === "CommonLawOffenceCode") {
        offenceCode.CommonLawOffence = COMMON_LAWS
      }
    } else if (spiOffenceCode.startsWith(INDICTMENT)) {
      offenceCode.__type = "IndictmentOffenceCode"
      if (offenceCode.__type === "IndictmentOffenceCode") {
        offenceCode.Indictment = INDICTMENT
      }
    } else {
      offenceCode.__type = "NonMatchingOffenceCode"
      if (offenceCode.__type === "NonMatchingOffenceCode") {
        offenceCode.ActOrSource = spiOffenceCodeLength < 2 ? spiOffenceCode : spiOffenceCode.substring(0, 2)

        offenceCode.Year =
          spiOffenceCodeLength > 2 ? spiOffenceCode.substring(2, Math.min(4, spiOffenceCodeLength)) : ""
      }
    }

    return offenceCode as OffenceCode
  }

  private populateOffence = (spiOffence: OffenceParsedXml): Offence | undefined => {
    const offence = {} as Offence
    const {
      BaseOffenceDetails: {
        OffenceCode: spiOffenceCode,
        OffenceSequenceNumber: spiOffenceSequenceNumber,
        ArrestDate: spiArrestDate,
        ChargeDate: spiChargeDate,
        LocationOfOffence: spiLocationOfOffence,
        OffenceWording: spiOffenceWording,
        AlcoholRelatedOffence: spiAlcoholRelatedOffence,
        OffenceTiming: { OffenceDateCode: spiOffenceDateCode, OffenceStart: spiOffenceStart, OffenceEnd: spiOffenceEnd }
      },
      ConvictionDate: spiConvictionDate,
      Result: spiResults
    } = spiOffence

    const enteredInError = spiResults.some(
      (result) => result.ResultCode !== undefined && Number(result.ResultCode) === enteredInErrorResultCode
    )
    if (enteredInError) {
      return undefined
    }

    const OffenceCode = this.getOffenceReason(spiOffenceCode ?? "")
    offence.CriminalProsecutionReference = {
      OffenceReason: {
        OffenceCode,
        __type: "NationalOffenceReason"
      }
    } as CriminalProsecutionReference

    offence.ArrestDate = spiArrestDate ? new Date(spiArrestDate) : undefined
    offence.ChargeDate = spiChargeDate ? new Date(spiChargeDate) : undefined
    offence.ActualOffenceDateCode = spiOffenceDateCode?.toString()
    offence.ActualOffenceStartDate = {
      StartDate: new Date(spiOffenceStart.OffenceDateStartDate)
    }
    offence.ActualOffenceEndDate = spiOffenceEnd ? { EndDate: new Date(spiOffenceEnd.OffenceEndDate) } : undefined
    offence.LocationOfOffence = spiLocationOfOffence
    offence.ActualOffenceWording = spiOffenceWording

    if (spiAlcoholRelatedOffence) {
      offence.AlcoholLevel = {
        Amount: Number(spiAlcoholRelatedOffence.AlcoholLevelAmount),
        Method:
          lookupAlcoholLevelMethodBySpiCode(spiAlcoholRelatedOffence.AlcoholLevelMethod)?.cjsCode ??
          spiAlcoholRelatedOffence.AlcoholLevelMethod
      }
    }

    if (spiOffenceStart?.OffenceStartTime) {
      const spiOffenceStartTime = removeSeconds(spiOffenceStart.OffenceStartTime)

      const offenceDateCode = Number(spiOffenceDateCode)
      const { ON_OR_IN, BEFORE, AFTER, ON_OR_ABOUT, ON_OR_BEFORE, BETWEEN } = timeRange
      if ([ON_OR_IN, BEFORE, AFTER, ON_OR_ABOUT, ON_OR_BEFORE].includes(offenceDateCode)) {
        offence.OffenceTime = spiOffenceStartTime
      } else if (offenceDateCode === BETWEEN) {
        offence.StartTime = spiOffenceStartTime
      }
    }

    if (spiOffenceEnd?.OffenceEndTime) {
      offence.OffenceEndTime = removeSeconds(spiOffenceEnd.OffenceEndTime)
    }

    offence.CommittedOnBail = dontKnowValue
    offence.CourtOffenceSequenceNumber = Number(spiOffenceSequenceNumber)

    offence.ConvictionDate = spiConvictionDate ? new Date(spiConvictionDate) : undefined

    if (!spiConvictionDate) {
      this.isAdjournmentSineDieConditionMet ||= adjournmentSineDieConditionMet(spiResults)

      if (this.isAdjournmentSineDieConditionMet) {
        offence.ConvictionDate = new Date(this.courtResult.Session.CourtHearing.Hearing.DateOfHearing)
      }
    }

    const { results, bailQualifiers } = new PopulateOffenceResults(this.courtResult, spiOffence).execute()
    offence.Result = results

    if (this.hearingDefendantBailConditions.length > 0 && bailQualifiers.length > 0) {
      bailQualifiers.forEach((bailQualifier) => {
        const description = lookupResultQualifierCodeByCjsCode(bailQualifier)?.description
        if (description) {
          this.bailConditions.push(description)
        }
      })
    }

    return offence
  }

  execute(): OffencesResult {
    const spiOffences = this.courtResult.Session.Case.Defendant.Offence
    const offences = spiOffences.map(this.populateOffence).filter((offence) => !!offence) as Offence[]
    this.bailConditions = this.hearingDefendantBailConditions.concat(this.bailConditions)

    return { offences, bailConditions: this.bailConditions }
  }
}
