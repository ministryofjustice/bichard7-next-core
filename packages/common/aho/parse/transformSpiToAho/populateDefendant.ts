import type { Address, DefendantDetail, HearingDefendant } from "../../../types/AnnotatedHearingOutcome"
import type {
  ResultedCaseMessageParsedXml,
  SpiAddress,
  SpiCourtIndividualDefendant,
  SpiDefendant
} from "../../../types/SpiResult"

import { lookupRemandStatusBySpiCode } from "../../dataLookup/dataLookup"
import populateOffences from "./populateOffences"

const formatPncIdentifier = (spiPNCIdentifier?: string): string | undefined =>
  spiPNCIdentifier ? spiPNCIdentifier.substring(0, 4) + "/" + spiPNCIdentifier.substring(4) : undefined

const populatePersonDefendantDetail = (spiCourtIndividualDefendant: SpiCourtIndividualDefendant): DefendantDetail => {
  const {
    PersonDefendant: {
      BasePersonDetails: {
        Birthdate: spiBirthDate,
        Gender: spiGender,
        PersonName: {
          PersonFamilyName: spiPersonFamilyName,
          PersonGivenName1: spiPersonGivenName1,
          PersonGivenName2: spiPersonGivenName2,
          PersonGivenName3: spiPersonGivenName3,
          PersonTitle: spiPersonTitle
        }
      }
    }
  } = spiCourtIndividualDefendant

  const spiGivenNames = [spiPersonGivenName1, spiPersonGivenName2, spiPersonGivenName3]
    .filter((name) => name !== undefined && name.trim() !== "")
    .map((name) => name!.trim())
    .map((name) => name.replace(/\s+/g, " "))

  return {
    BirthDate: spiBirthDate ? new Date(spiBirthDate) : undefined,
    Gender: Number(spiGender),
    PersonName: {
      FamilyName: spiPersonFamilyName.replace(/\s+/g, " ").trim(),
      GivenName: spiGivenNames,
      Title: spiPersonTitle?.trim() || undefined
    }
  }
}

const populateAddress = (spiAddress: SpiAddress): Address => {
  if ("SimpleAddress" in spiAddress) {
    const {
      SimpleAddress: { AddressLine1, AddressLine2, AddressLine3, AddressLine4, AddressLine5 }
    } = spiAddress

    return {
      AddressLine1: AddressLine1,
      AddressLine2: AddressLine2,
      AddressLine3: AddressLine3,
      AddressLine4: AddressLine4,
      AddressLine5: AddressLine5
    }
  } else {
    const {
      ComplexAddress: { AdministrativeArea, Locality, StreetDescription, Town, UniqueStreetReferenceNumber }
    } = spiAddress

    const hearingOutcomeAddress = [
      UniqueStreetReferenceNumber?.toString(),
      StreetDescription,
      Locality,
      Town,
      AdministrativeArea
    ]
      .filter((x) => !!x)
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      .reduce((acc: any, addressPart, index) => {
        acc[`AddressLine${index + 1}`] = addressPart!
        return acc
      }, {})

    return hearingOutcomeAddress
  }
}

const parseBailConditions = (spiBailConditions?: string): string[] => {
  if (spiBailConditions && spiBailConditions !== "") {
    return spiBailConditions
      .split(";")
      .map((cond) => cond.trim())
      .filter((cond) => !!cond)
  }

  return []
}

type PartialDefendantDetails =
  | Pick<HearingDefendant, "Address" | "BailConditions" | "OrganisationName" | "PNCIdentifier" | "RemandStatus">
  | Pick<
      HearingDefendant,
      | "Address"
      | "BailConditions"
      | "CourtPNCIdentifier"
      | "DefendantDetail"
      | "ReasonForBailConditions"
      | "RemandStatus"
    >

const processDefendant = (spiDefendant: SpiDefendant): PartialDefendantDetails => {
  if (spiDefendant.CourtIndividualDefendant) {
    const {
      CourtIndividualDefendant: {
        Address: spiAddress,
        BailStatus: spiBailStatus,
        PersonDefendant: { BailConditions: spiBailConditions, PNCidentifier: spiPNCidentifier },
        ReasonForBailConditionsOrCustody: spiReasonForBailConditionsOrCustody
      }
    } = spiDefendant

    return {
      Address: populateAddress(spiAddress),
      BailConditions: parseBailConditions(spiBailConditions),
      CourtPNCIdentifier: formatPncIdentifier(spiPNCidentifier),
      DefendantDetail: populatePersonDefendantDetail(spiDefendant.CourtIndividualDefendant),
      ReasonForBailConditions: spiReasonForBailConditionsOrCustody,
      RemandStatus: lookupRemandStatusBySpiCode(spiBailStatus)?.cjsCode ?? spiBailStatus
    }
  }

  if (spiDefendant.CourtCorporateDefendant) {
    // Corporate Defendant
    const {
      Address: spiAddress,
      BailStatus: spiBailStatus,
      OrganisationName: { OrganisationName: spiOrganisationName },
      PNCidentifier: spiPNCidentifier
    } = spiDefendant.CourtCorporateDefendant

    return {
      Address: populateAddress(spiAddress),
      BailConditions: [],
      CourtPNCIdentifier: formatPncIdentifier(spiPNCidentifier),
      OrganisationName: spiOrganisationName,
      RemandStatus: lookupRemandStatusBySpiCode(spiBailStatus)?.cjsCode ?? spiBailStatus
    }
  }

  throw new Error("CourtIndividualDefendant and CourtCorporateDefendant are missing from incoming record")
}

export default (courtResult: ResultedCaseMessageParsedXml): HearingDefendant => {
  const {
    Session: {
      Case: { Defendant: spiDefendant }
    }
  } = courtResult

  const { bailConditions, offences } = populateOffences(courtResult)

  const partialDefendant = processDefendant(spiDefendant)

  if (partialDefendant.BailConditions.length > 0) {
    partialDefendant.BailConditions = partialDefendant.BailConditions.concat(bailConditions)
  }

  const hearingDefendant: HearingDefendant = {
    ...partialDefendant,
    ArrestSummonsNumber: spiDefendant.ProsecutorReference,
    Offence: offences
  }

  return hearingDefendant
}
