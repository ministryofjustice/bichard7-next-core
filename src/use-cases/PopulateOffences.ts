import {
  ADJOURNMENT_SINE_DIE_RESULT_CODE_STRING,
  COMMON_LAWS,
  DONT_KNOW_VALUE,
  ENTERED_IN_ERROR_RESULT_CODE,
  INDICTMENT,
  STOP_LIST,
  TIME_RANGE
} from "src/lib/properties"
import type { Offence, OffenceCode } from "src/types/AnnotatedHearingOutcome"
import type { OffenceParsedXml, ResultedCaseMessageParsedXml } from "src/types/IncomingMessage"
import removeSeconds from "src/utils/removeSeconds"
import {
  lookupAlcoholLevelMethodBySpiCode,
  lookupOffenceCodeByCjsCode,
  lookupResultQualifierCodeByCjsCode
} from "./dataLookup"

import PopulateOffenceResults from "./PopulateOffenceResults"

export interface OffencesResult {
  offences: Offence[]
  bailConditions: string[]
}

export default class {
  private adjournmentSineDieConditionMet = false

  private bailConditions: string[] = []

  constructor(private courtResult: ResultedCaseMessageParsedXml, private hearingDefendantBailConditions: string[]) {}

  private getOffenceReason = (spiOffenceCode: string): OffenceCode => {
    const spiOffenceCodeLength = spiOffenceCode.length
    const offenceCode: Partial<OffenceCode> = {
      Reason: spiOffenceCodeLength > 4 ? spiOffenceCode.substring(4, Math.min(7, spiOffenceCodeLength)) : "",
      ...(spiOffenceCodeLength > 7 ? { Qualifier: spiOffenceCode.substring(7) } : {}),
      FullCode: spiOffenceCode
    }

    if (spiOffenceCode.startsWith(COMMON_LAWS)) {
      offenceCode.__type = "CommonLawOffenceCode"
      if (offenceCode.__type === "CommonLawOffenceCode") {
        offenceCode.CommonLawOffence = COMMON_LAWS
      }
    } else if (spiOffenceCode.startsWith(INDICTMENT)) {
      offenceCode.__type = "IndictmentOffenceCode"
      if (offenceCode.__type === "IndictmentOffenceCode") {
        offenceCode.Indictment = INDICTMENT
      }
    } else {
      offenceCode.__type = "NonMatchingOffenceCode"
      if (offenceCode.__type === "NonMatchingOffenceCode") {
        offenceCode.ActOrSource = spiOffenceCodeLength < 2 ? spiOffenceCode : spiOffenceCode.substring(0, 2)

        offenceCode.Year =
          spiOffenceCodeLength > 2 ? spiOffenceCode.substring(2, Math.min(4, spiOffenceCodeLength)) : ""
      }
    }

    return offenceCode as OffenceCode
  }

  private populateOffence = (spiOffence: OffenceParsedXml): Offence | undefined => {
    const offence = {} as Offence
    const {
      BaseOffenceDetails: {
        OffenceCode: spiOffenceCode,
        OffenceSequenceNumber: spiOffenceSequenceNumber,
        ArrestDate: spiArrestDate,
        ChargeDate: spiChargeDate,
        LocationOfOffence: spiLocationOfOffence,
        OffenceWording: spiOffenceWording,
        AlcoholRelatedOffence: spiAlcoholRelatedOffence,
        OffenceTiming: { OffenceDateCode: spiOffenceDateCode, OffenceStart: spiOffenceStart, OffenceEnd: spiOffenceEnd }
      },
      ConvictionDate: spiConvictionDate,
      Result: spiResults
    } = spiOffence

    spiResults.forEach((spiResult) => {
      if (spiResult.ResultCode === ENTERED_IN_ERROR_RESULT_CODE) {
        console.log(`Offence entered in error: ${spiOffenceCode}`)
        return undefined
      }
    })

    const OffenceCode = this.getOffenceReason(spiOffenceCode ?? "")
    offence.CriminalProsecutionReference = {
      OffenceReason: {
        OffenceCode,
        __type: "NationalOffenceReason"
      }
    }

    offence.ArrestDate = spiArrestDate ? new Date(spiArrestDate) : undefined
    offence.ChargeDate = spiChargeDate ? new Date(spiChargeDate) : undefined
    offence.ActualOffenceDateCode = spiOffenceDateCode?.toString()
    offence.ActualOffenceStartDate = {
      StartDate: new Date(spiOffenceStart.OffenceDateStartDate)
    }
    offence.ActualOffenceEndDate = {
      EndDate: spiOffenceEnd ? new Date(spiOffenceEnd.OffenceEndDate) : undefined
    }
    offence.LocationOfOffence = spiLocationOfOffence
    offence.ActualOffenceWording = spiOffenceWording

    if (spiAlcoholRelatedOffence) {
      offence.AlcoholLevel = {
        Amount: spiAlcoholRelatedOffence.AlcoholLevelAmount.toString(),
        Method:
          lookupAlcoholLevelMethodBySpiCode(spiAlcoholRelatedOffence.AlcoholLevelMethod)?.cjsCode ??
          spiAlcoholRelatedOffence.AlcoholLevelMethod
      }
    }

    if (spiOffenceStart?.OffenceStartTime) {
      const spiOffenceStartTime = removeSeconds(spiOffenceStart.OffenceStartTime)

      const { ON_OR_IN, BEFORE, AFTER, ON_OR_ABOUT, ON_OR_BEFORE, BETWEEN } = TIME_RANGE
      if ([ON_OR_IN, BEFORE, AFTER, ON_OR_ABOUT, ON_OR_BEFORE].includes(spiOffenceDateCode)) {
        offence.OffenceTime = spiOffenceStartTime
      } else if (spiOffenceDateCode === BETWEEN) {
        offence.StartTime = spiOffenceStartTime
      }
    }

    if (spiOffenceEnd?.OffenceEndTime) {
      offence.OffenceEndTime = removeSeconds(spiOffenceEnd.OffenceEndTime)
    }

    offence.ConvictionDate = spiConvictionDate ? new Date(spiConvictionDate) : undefined
    offence.CommittedOnBail = DONT_KNOW_VALUE
    offence.CourtOffenceSequenceNumber = spiOffenceSequenceNumber

    if (!spiConvictionDate) {
      const a2007ResultFound = spiResults.some(
        (spiResult) => spiResult.ResultCode?.toString() === ADJOURNMENT_SINE_DIE_RESULT_CODE_STRING
      )
      const aFailConditionResultFound = spiResults.some(
        (spiResult) => spiResult.ResultCode && STOP_LIST.includes(spiResult.ResultCode)
      )

      if (a2007ResultFound && !aFailConditionResultFound) {
        this.adjournmentSineDieConditionMet = true
      }

      if (this.adjournmentSineDieConditionMet) {
        offence.ConvictionDate = new Date(this.courtResult.Session.CourtHearing.Hearing.DateOfHearing)
      }
    }

    const { results, bailQualifiers } = new PopulateOffenceResults(this.courtResult, spiOffence).execute()
    offence.Result = results

    if (this.hearingDefendantBailConditions.length > 0 && bailQualifiers.length > 0) {
      bailQualifiers.forEach((bailQualifier) => {
        const description = lookupResultQualifierCodeByCjsCode(bailQualifier)?.description
        if (description) {
          this.bailConditions.push(description)
        }
      })
    }

    const offenceCode = lookupOffenceCodeByCjsCode(spiOffenceCode)
    if (offenceCode) {
      offence.RecordableOnPNCindicator = offenceCode.recordableOnPnc === "Y"
    }

    return offence
  }

  execute(): OffencesResult {
    const spiOffences = this.courtResult.Session.Case.Defendant.Offence
    const offences = spiOffences.map(this.populateOffence).filter((offence) => !!offence) as Offence[]
    this.bailConditions = this.hearingDefendantBailConditions.concat(this.bailConditions)

    return { offences, bailConditions: this.bailConditions }
  }
}
