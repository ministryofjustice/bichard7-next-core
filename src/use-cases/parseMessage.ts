import { XMLParser } from "fast-xml-parser"
import type { ResultedCaseMessageParsedXml } from "src/types/IncomingMessage"
import { incomingMessageParsedXmlSchema } from "src/types/IncomingMessage"

export default (message: string): ResultedCaseMessageParsedXml => {
  const options = {
    ignoreAttributes: false,
    removeNSPrefix: true
  }

  const parser = new XMLParser(options)
  const rawParsedObj = parser.parse(message) as unknown
  const parsedObj = incomingMessageParsedXmlSchema.parse(rawParsedObj)
  const rcm = parsedObj.DeliverRequest.Message.ResultedCaseMessage
  // if (!rcm.Session.Case.Defendant.Offence) {
  //   rcm.Session.Case.Defendant.Offence = []
  // }
  // if (!Array.isArray(rcm.Session.Case.Defendant.Offence)) {
  //   rcm.Session.Case.Defendant.Offence = [rcm.Session.Case.Defendant.Offence]
  // }
  // rcm.Session.Case.Defendant.Offence.forEach((offence) => {
  //   if (!offence.Result) {
  //     offence.Result = []
  //   }
  //   if (!Array.isArray(offence.Result)) {
  //     offence.Result = [offence.Result]
  //   }
  // })
  return rcm
}
