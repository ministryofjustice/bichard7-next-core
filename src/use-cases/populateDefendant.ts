import type { Address, DefendantDetail, HearingDefendant } from "src/types/HearingOutcome"
import type { ResultedCaseMessageParsedXml, SpiAddress, SpiCourtIndividualDefendant } from "src/types/IncomingMessage"
import type { ErrorAttribute } from "src/types/XmlElement"
import { createElement } from "src/types/XmlElement"
import PopulateOffences from "./PopulateOffences"

const populatePersonDefendantDetail = (spiCourtIndividualDefendant: SpiCourtIndividualDefendant): DefendantDetail => {
  const {
    PersonDefendant: {
      BasePersonDetails: {
        PersonName: {
          PersonTitle: spiPersonTitle,
          PersonGivenName1: spiPersonGivenName1,
          PersonGivenName2: spiPersonGivenName2,
          PersonGivenName3: spiPersonGivenName3,
          PersonFamilyName: spiPersonFamilyName
        },
        Gender: spiGender,
        Birthdate: spiBirthDate
      }
    }
  } = spiCourtIndividualDefendant

  const spiGivenNames = [spiPersonGivenName1, spiPersonGivenName2, spiPersonGivenName3].filter((name) => !!name)

  return {
    PersonName: {
      Title: createElement(spiPersonTitle?.trim()),
      GivenName: spiGivenNames.map((name, index) =>
        createElement((name as string).trim(), { NameSequence: `${index + 1}` })
      ),
      FamilyName: createElement(spiPersonFamilyName, { NameSequence: "1" })
    },
    BirthDate: createElement(spiBirthDate),
    Gender: createElement(spiGender.toString())
  }
}

const populateAddress = (spiAddress: SpiAddress): Address | undefined => {
  if ("SimpleAddress" in spiAddress) {
    const {
      SimpleAddress: { AddressLine1, AddressLine2, AddressLine3, AddressLine4, AddressLine5 }
    } = spiAddress

    return {
      AddressLine1: createElement(AddressLine1),
      AddressLine2: createElement(AddressLine2),
      AddressLine3: createElement(AddressLine3),
      AddressLine4: createElement(AddressLine4),
      AddressLine5: createElement(AddressLine5)
    }
  } else if ("ComplexAddress" in spiAddress) {
    const {
      ComplexAddress: { Locality, StreetDescription, Town, AdministrativeArea, UniqueStreetReferenceNumber }
    } = spiAddress

    const hearingOutcomeAddress = [
      UniqueStreetReferenceNumber?.toString(),
      StreetDescription,
      Locality,
      Town,
      AdministrativeArea
    ]
      .filter((x) => !!x)
      .reduce((acc: any, addressPart, index) => {
        acc[`AddressLine${index + 1}`] = createElement<string, ErrorAttribute>(addressPart!)
        return acc
      }, {})

    return {
      ...hearingOutcomeAddress
    }
  }

  return undefined
}

const populateRemandStatus = (spiBailStatus: string): string => {
  // RemandStatus <- BailStatus ?
  return `${spiBailStatus}: RemandStatus <- BailStatus ?`
}

export default (courtResult: ResultedCaseMessageParsedXml): HearingDefendant => {
  const hearingDefendant = {} as HearingDefendant
  const {
    Session: {
      Case: { Defendant: spiDefendant }
    }
  } = courtResult

  hearingDefendant.ArrestSummonsNumber = createElement(spiDefendant.ProsecutorReference)

  if (spiDefendant.CourtIndividualDefendant) {
    const {
      CourtIndividualDefendant: { PersonDefendant: spiPersonDefendant, BailStatus: spiBailStatus, Address: spiAddress }
    } = spiDefendant
    hearingDefendant.CourtPNCIdentifier = createElement(spiPersonDefendant.PNCidentifier)
    hearingDefendant.DefendantDetail = populatePersonDefendantDetail(spiDefendant.CourtIndividualDefendant)
    hearingDefendant.Address = populateAddress(spiAddress)

    // RemandStatus <- BailStatus ?
    hearingDefendant.RemandStatus = createElement(populateRemandStatus(spiBailStatus))

    hearingDefendant.BailConditions =
      spiPersonDefendant.BailConditions?.split(";").map((bailCondition) => createElement(bailCondition)) || []

    hearingDefendant.ReasonForBailConditions = createElement(spiPersonDefendant.ReasonForBailConditionsOrCustody)
  } else if (spiDefendant.CourtCorporateDefendant) {
    // Corporate Defendant
    const {
      PNCidentifier: spiPNCidentifier,
      OrganisationName: { OrganisationName: spiOrganisationName },
      Address: spiAddress,
      BailStatus: spiBailStatus
    } = spiDefendant.CourtCorporateDefendant

    hearingDefendant.CourtPNCIdentifier = createElement(spiPNCidentifier)
    hearingDefendant.OrganisationName = createElement(spiOrganisationName)
    hearingDefendant.Address = populateAddress(spiAddress)

    // RemandStatus <- BailStatus ?
    hearingDefendant.RemandStatus = createElement(populateRemandStatus(spiBailStatus))

    const spiDateOfHearing = courtResult.Session.CourtHearing.Hearing.DateOfHearing
    hearingDefendant.Offence = new PopulateOffences(spiDefendant.Offence, spiDateOfHearing).execute()
  }

  return hearingDefendant
}
