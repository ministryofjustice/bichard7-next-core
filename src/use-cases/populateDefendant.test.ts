import { readFileSync } from "fs"
import type { Offence } from "src/types/HearingOutcome"
import parseMessage from "./parseMessage"
import populateDefendant from "./populateDefendant"
import PopulateOffences from "./PopulateOffences"

describe("populateDefendant", () => {
  const message = readFileSync("test-data/input-message-001.xml", "utf-8")
  const courtResult = parseMessage(message)

  it("should transform SPI Defendant to Hearing Outcome Defendant for individual defendant", () => {
    const result = populateDefendant(courtResult)

    jest.spyOn(PopulateOffences.prototype, "execute").mockReturnValue({} as Offence[])
    expect(result).toBeDefined()
    expect(result).toMatchSnapshot()
  })
})
