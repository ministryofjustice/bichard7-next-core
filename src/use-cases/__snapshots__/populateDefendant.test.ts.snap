// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`populateDefendant should transform SPI Defendant to Hearing Outcome Defendant for individual defendant 1`] = `
Object {
  "Address": Object {
    "AddressLine1": Object {
      "attributes": undefined,
      "value": "Scenario1 Address Line 1",
    },
    "AddressLine2": Object {
      "attributes": undefined,
      "value": "Scenario1 Address Line 2",
    },
    "AddressLine3": Object {
      "attributes": undefined,
      "value": "Scenario1 Address Line 3",
    },
    "AddressLine4": Object {
      "attributes": undefined,
      "value": undefined,
    },
    "AddressLine5": Object {
      "attributes": undefined,
      "value": undefined,
    },
  },
  "ArrestSummonsNumber": Object {
    "attributes": undefined,
    "value": "1101ZD0100000448754K",
  },
  "BailConditions": Array [],
  "CourtPNCIdentifier": Object {
    "attributes": undefined,
    "value": undefined,
  },
  "DefendantDetail": Object {
    "BirthDate": Object {
      "attributes": undefined,
      "value": "1948-11-11",
    },
    "Gender": Object {
      "attributes": undefined,
      "value": "1",
    },
    "PersonName": Object {
      "FamilyName": Object {
        "attributes": Object {
          "NameSequence": "1",
        },
        "value": "SEXOFFENCE",
      },
      "GivenName": Array [
        Object {
          "attributes": Object {
            "NameSequence": "1",
          },
          "value": "TRPRFOUR",
        },
      ],
      "Title": Object {
        "attributes": undefined,
        "value": "Mr",
      },
    },
  },
  "ReasonForBailConditions": Object {
    "attributes": undefined,
    "value": undefined,
  },
  "RemandStatus": Object {
    "attributes": undefined,
    "value": "U: RemandStatus <- BailStatus ?",
  },
}
`;
